mava/systems/q_learning/sebulba/rec_iql.py

```diff
@@ -0,0 +1,782 @@
+                        dones,
+                        next_timestep.extras["real_next_obs"],
+                    )
+                )
+
+                episode_metrics.append(timestep.extras["episode_metrics"])
+
+        # send trajectories to learner
+        with RecordTimeTo(actor_timings["rollout_put_time"]):
+            try:
+                rollout_queue.put(traj, (actor_timings, episode_metrics), actor_id)
+            except queue.Full:
+                err = "Waited too long to add to the rollout queue, killing the actor thread"
+                warnings.warn(err, stacklevel=2)
+                break
+
+    env.close()
+
+
+def get_learner_step_fn(
+    q_net: RecQNetwork,
+    update_fn: optax.TransformUpdateFn,
+    config: DictConfig,
+) -> SebulbaLearnerFn[LearnerState, Transition]:
+    """Get the learner function."""
+
+    def _update_step(
+        learner_state: LearnerState,
+        traj_batch: Transition,
+    ) -> Tuple[LearnerState, Metrics]:
+        """Performs a single network update.
+
+        Calculates targets based on the input trajectories and updates the Q-network
+        parameters accordingly.
+
+        Args:
+            learner_state: Current learner state.
+            traj_batch: Batch of transitions for training.
+        """
+
+        def prep_inputs_to_scannedrnn(obs: Observation, term_or_trunc: chex.Array) -> chex.Array:
+            """Prepares inputs for the ScannedRNN network.
+
+            Switches leading axes of observations and termination/truncation flags to match the
+            (T, B, ...) format expected by the RNN. The replay buffer outputs data in (B, T, ...)
+            format.
+
+            Args:
+                obs: Observation data.
+                term_or_trunc: Termination/truncation flags.
+
+            Returns:
+                Tuple containing the initial hidden state and the formatted input data.
+            """
+            hidden_state = ScannedRNN.initialize_carry(
+                (obs.agents_view.shape[0], obs.agents_view.shape[2]),
+                config.network.hidden_state_dim,
+            )
+            # the rb outputs (B, T, ... ) the RNN takes in (T, B, ...)
+            obs = switch_leading_axes(obs)  # (B, T) -> (T, B)
+            term_or_trunc = switch_leading_axes(term_or_trunc)  # (B, T) -> (T, B)
+            obs_term_or_trunc = (obs, term_or_trunc)
+
+            return hidden_state, obs_term_or_trunc
+
+        def _update_epoch(update_state: Tuple, _: Any) -> Tuple:
+            """Update the network for a single epoch."""
+
+            def q_loss_fn(
+                q_online_params: FrozenVariableDict,
+                obs: Array,
+                term_or_trunc: Array,
+                action: Array,
+                target: Array,
+            ) -> Tuple[Array, Metrics]:
+                # axes switched here to scan over time
+                hidden_state, obs_term_or_trunc = prep_inputs_to_scannedrnn(obs, term_or_trunc)
```

Comment:
A general comment, I think this would be a lot easier to read if we used `done` to mean `term_or_trunc` which I think is a reasonable thing. Would have to make the change in anakin also though :/